sycl/doc/design/DeviceIf.md

```diff
@@ -338,4 +338,1011 @@ in that compilation phase.
+    * The set of aspects used by the _Action_ function and its static call tree
+      (i.e. from the `!sycl_used_aspects` metadata).
+    * A generation number (described below).
+
+  * If _Action_ is not in the table and if the device image does not already
+    contain the function _Action_, add it to the table with _Condition_ and set
+    the generation number to 1.  (The only time when the device image already
+    contains _Action_ is when there is some call to _Action_ other than from a
+    _Conditional Caller_ function.)
+
+  * If _Action_ is in the table with the same _Condition_, get the generation
+    number from that table entry.
+
+  * Otherwise, one of two weird things has happened.  Either there are two (or
+    more) calls to _Action_ with different conditions, or there is also a call
+    to _Action_ from someplace other than a _Conditional Caller_.  Both cases
+    are handled the same way.  Choose a generation number such that the name
+    @CallableXXX.GENERATION is globally unique.  Add a new entry to the table
+    with _Condition_ and that generation number.  Duplicate the function
+    @CallableXXX as @CallableXXX.GENERATION.
+
+  * Change the call to this (if the generation number is 1):
+
+    ```
+    call void @CallableXXX(%callablethis)
+    ```
+
+    or to this (if the generation number is something else):
+
+    ```
+    call void @CallableXXX.GENERATION(%callablethis)
+    ```
+
+  * Remember that the device code split algorithm did not include the
+    definition of the _Action_ function in the image.  Therefore, this function
+    is declared (not defined) with external linkage.
+
+* The LLVM IR bitcode for this device image is now ready to be emitted.  We
+  call this the _Main Image_.
+
+* Using the association table described earlier, separate the _Conditional
+  Action_ functions into groups such that all functions in the same group share
+  the same _Condition_ and the same set of used aspects.  For each such group:
+
+  * Construct a new device image that contains the definition of each
+    _Conditional Action_ function from the group along with the static call
+    tree of those functions.  This code may make additional calls to the
+    _Conditional Caller_ functions, passing pointers to yet more _Conditional
+    Action_ functions.  Do not follow these function pointers when constructing
+    the static call tree.
+
+  * Construct another new device image that contains stub definitions of each
+    _Conditional Action_ function in the group.  Each stub defines the function
+    with an empty body.
+
+  * The definitions of the the _Conditional Action_ functions in these two
+    images must have external linkage.  We call these new device images the
+    _Add On Images_.
+
+  * Each pair of _Add On Images_ (i.e. the one with the real function
+    definitions and the one with the stub definitions) is assigned a unique
+    name.  By convention this is just an integer in string form (e.g. "1").
```

Comment:
I was thinking that the "scope" of the _Add On Image_ numbers was limited to the ELF file that contains them.  For example, if a _Main Image_ contains a reference to _Add On Image_ "1", we only search for that _Add On Image_ in the same ELF file.

If you think this will be difficult and we need a globally unique name, we could use a GUID.  However, I think the driver will need to pass a GUID via a command line option to `sycl-post-link`.